### ISSUE TITLE
apf::Mesh::getParam for Quadratic Meshes

### ISSUE THREAD

**Issue by JaredCrean2** [SCOREC/core#213]
What is the status of [getParam](https://scorec.rpi.edu/~seol/scorec/doxygen/classapf_1_1Mesh.html#afa3c34a178871cfb9718271085c7359a) for quadratic meshes? It does not take a `node` argument like `getPoint` does, and I am seeing what looks like uninitialized memory sometimes when I call it on mesh edges.  The full code is in `/users/creanj/test_quadratic`, but the main part is:

```
void printParams(apf::Mesh* m)
{

  apf::FieldShape* fshape = m->getShape();
  std::cout << "Printing CAD parametric coordinates for " << fshape->getName() << std::endl;

  apf::MeshEntity* e;
  apf::MeshIterator* it;
  apf::Vector3 p; p[2] = 0;
  apf::ModelEntity* me;
  for (int dim=0; dim <= m->getDimension(); ++dim)
  {
    if (!fshape->hasNodesIn(dim))
      continue;

    std::cout << "Dimension " << dim << std::endl;
    it = m->begin(dim);
    while ( (e = m->iterate(it)) )
    {
      // assume 1 node per entity
      m->getParam(e, p);
      me = m->toModel(e);
      int tag = m->getModelTag(me);
      int me_dim = m->getModelType(me);
      std::cout << "  entity " << e << " classified on model entity " << tag << " of dimension " << me_dim << " has parametric coords = " << p.x();
      std::cout << ", " << p.y() << ", " << p.z() << std::endl;

    }
    m->end(it);

  }
}
```
For mesh edges, an excerpt of the output is:
```
  entity 0x6f2 classified on model entity 1 of dimension 2 has parametric coords = 7.63564e-317, 7.63564e-317, 0
  entity 0x6fa classified on model entity 1 of dimension 2 has parametric coords = 7.63564e-317, 7.63564e-317, 0
  entity 0x702 classified on model entity 1 of dimension 2 has parametric coords = 7.63564e-317, 7.63564e-317, 0
  entity 0x70a classified on model entity 1 of dimension 2 has parametric coords = 7.63564e-317, 7.63564e-317, 0
  entity 0x712 classified on model entity 1 of dimension 2 has parametric coords = 7.63564e-317, 7.63564e-317, 0
```

**Comment by mortezah** [comment]
@JaredCrean2, that information is only stored (an updated if need be) for mesh vertices any other node (like edge nodes) will not have it. We actually do not need it during mesh adapt or curved mesh adapt. So it makes sense that you get uninitialized values for them.

Is there any specific situation that you need that information for?

**Comment by JaredCrean2** [comment]
I'm doing optimization involving the mesh, so what I need to do is compute derivatives like `df/dxi = df/dx * dxi/dx`, where `xi` is a CAD parametric coordinate.  The `dxi/dx` needs to be evaluated at the parametric coordinate of the edge node.

**Comment by mortezah** [comment]
here are two things that you could do

1) use `apf::Mesh2::getClosestPoint (ModelEntity *g, Vector3 const &from, Vector3 &to, Vector3 &p)` where `from` is the location of the node as obtained by `getPoint`. Then `to` should be the same as `from` and `p` will be the parametric coordinates on the corresponding model entity `g`.

Then you can use `apf::Mesh2::getFirstDerivative (ModelEntity *g, Vector3 const &p, Vector3 &t0, Vector3 &t1)` to get the derivatives (note that `t0` and `t1` are kind of the inverse of what you are looking for, which is dxi/dx).

2) You could use `crv::transferParametricOnEdgeSplit (apf::Mesh *m, apf::MeshEntity *e, double t, apf::Vector3 &p)` or `transferParametricOnTriSplit (apf::Mesh2 *m, apf::MeshEntity *e, apf::Vector3 &t, apf::Vector3 &p)`, depending on the entity type. Thes routines interpolate the parametric location in the corresponding mesh entities, based on the parametric locations of the downward adjacent edges. 

`apf::Mesh2::getFirstDerivative (ModelEntity *g, Vector3 const &p, Vector3 &t0, Vector3 &t1)` can be used as before, afterwards.
 
The second approach is supposed to be faster.

**Comment by JaredCrean2** [comment]
Does the mesh data structure allocate space to store the edge node parametric coordinates, and they are never set, or is the space not allocated at all?  I ask because, in order to get better accuracy, I will have to
  * store the output of `af::Mesh2::getClosestPoint` to a field (the parametric coordinates computed by `getClosestPoint` are not quite accurate enough to compute derivatives to the precision I need).
  * snap the edge nodes to the positions computes from the parametric coordinates (to ensure consistency)
  * write wrappers around `getParam/setParam` to provide a uniform interface for vertices and edges

This would be a bit easier if I can use the existing infrastructure (and it shouldn't be that different from storing the quadratic coordinate field).

The other consideration is that it would be better to get the mid edge coordinates directly from the original source (the Simmetrix mesh), rather than doing an approximate snapping procedure, so it might make sense to build this into the conversion function for Simmetrix meshes to Scorec meshes.

**Comment by mortezah** [comment]
To answer your first questions, as you can see here https://github.com/SCOREC/core/blob/564a703cedc77f45693f46fadaadbb6ae9a3f566/mds/mds_apf.c#L27
in mds we only allocate space for parameters of the verts.

- I don't understand your point about accuracy. 
- The edge nodes are already snapped to the model, so you don't have to do anything.

**Comment by JaredCrean2** [comment]
Ok, looks like I will have to manage the storage myself.

The accuracy issue is that if you do
```
apf::ModelEntity* me = m->toModel(e);  // e is some apf::MeshEntity* vertex
apf::Vector3 x, xnew, p, p2;
m->getPoint(e, 0, x);
m->getClosestPoint(me, x, xnew, p);
m->getParam(e, p2)
```
then `p - p2` is large enough that computing `dxi/dx` using `p` rather than `p2`.  result in inaccurate derivatives.  This is why I want to be careful about how the `p` is obtained.  For edge nodes, the best case is to get the cartesian and parametric coordinates directly from the mesh generator (which should guarantee they are consistent).  The second best option is to use `getClosestPoint` to find parametric coordinates, and then use `gmi_eval` to convert them to cartesian (`gmi_eval` is much more accurate than `getClosestPoint`).

If MDS doesn't store the parametric coordinates for edge nodes, I think the solution (for my code) is to write a replacement for `convert` that adds the Field with the edge node parametric coordinates, and write the wrappers around `setParam/getParam`.

**Comment by mortezah** [comment]
`p - p2` should be zero within the machine precision. Do you actually see non-zero values for the difference? Is this for a mesh that is just loaded or for a mesh that is adapted or curved? If yes, can you send me the mesh/model example along with the code that you are using to get the values?

gmi_eval and getClosestPoint (which is basically gmi_get_closest_point underneath) both use the CAD engine, so I don't think they are different in terms of accuracy.

**Comment by JaredCrean2** [comment]
> `p - p2` should be zero within the machine precision for vertices

No, I filed a ticket with Simmetrix support about this.  It turns out it does an iterative search of some kind and terminates based on some tolerance.  I found a test case where the true xi coordinate is 2.6655562599818e-6 but the one returned by `gmi_closest_point` is 2.6304000000000011e-06.  Errors in the second significant digit can be enough to cause problems, even though the magnitude of the error is small.

 I had to write the test case using Simmetrix APIs, but it can be found in `/users/creanj/fasttmp2/test_simmetrix`  The coordinate transformation stuff is around line 117.  The mesh came directly out of SimModeler, and has not been adapted.
